File: frameworks/rails/app/controllers/benchmark_controller.rb
# frozen_string_literal: true

require 'zlib'
require 'pg'

class BenchmarkController < ActionController::API
  mattr_accessor :dataset

  DATA_DIR = ENV.fetch('DATA_DIR', '/data')
  dataset_path = File.join(DATA_DIR, 'dataset.json')
  static_dir = File.join(DATA_DIR, 'static')

  if File.exist?(dataset_path)
    self.dataset = JSON.parse(File.read(dataset_path)).freeze
  end

  FileUtils.cp_r(File.join(DATA_DIR, 'static'), Rails.root.join('public', 'static'))

  PG_QUERY = 'SELECT id, name, category, price, quantity, active, tags, rating_score, rating_count FROM items WHERE price BETWEEN $1 AND $2 LIMIT $3'.freeze

  def pipeline
    render plain: 'ok'
  end

  def baseline11
    total = params[:a].to_i + params[:b].to_i
    if request.post?
      total += request.body.read.to_i
    end
    render plain: total.to_s
  end

  def baseline2
    total = params[:a].to_i + params[:b].to_i
    render plain: total.to_s
  end

  def json_endpoint
    return head(500) unless dataset

    m = (params[:m] || 1).to_i
    count = params[:count].to_i
    items = dataset.slice(0, count).map do |d|
      d.merge('total' => d['price'] * d['quantity'] * m)
    end

    result = JSON.generate({ 'items' => items, 'count' => items.length })

    if accept_encodings = request.headers['Accept-Encoding']
      types = accept_encodings.split(',').map(&:strip)
      if types.include? 'gzip'
        sio = StringIO.new
        gz = Zlib::GzipWriter.new(sio, 1)
        gz.write(result)
        gz.close
        response.headers['content-type'] = 'application/json'
        response.headers['content-encoding'] = 'gzip'
        send_data sio.string, disposition: :inline
      else
        response.headers['content-type'] = 'application/json'
        render plain: result
      end
    else
      response.headers['content-type'] = 'application/json'
      render plain: result
    end
  end

  def async_db
    min_val = (params[:min] || 10).to_i
    max_val = (params[:max] || 50).to_i
    limit = (params['limit'] || 50).to_i.clamp(1, 50)

    rows = self.class.get_async_db&.with do |connection|
      connection.exec_prepared('select', [min_val, max_val, limit])
    end || []

    items = rows.map do |r|
      {
        'id' => r['id'],
        'name' => r['name'],
        'category' => r['category'],
        'price' => r['price'],
        'quantity' => r['quantity'],
        'active' => r['active'] == 't',
        'tags' => JSON.parse(r['tags']),
        'rating' => { 'score' => r['rating_score'], 'count' => r['rating_count'] }
      }
    end
    render json: { items: items, count: items.length }
  end

  def upload
    size = 0
    buf = request.body
    while (chunk = buf.read(65536))
      size += chunk.bytesize
    end
    render plain: size.to_s
  end

  def not_found
    head 404
  end

  private

  def self.get_async_db
    @async_db ||= begin
      return unless ENV['DATABASE_URL'].present?
      max_connections = ENV.fetch('RAILS_MAX_THREADS', 4).to_i
      ConnectionPool.new(size: max_connections, timeout: 5) do
        db = PG.connect(ENV['DATABASE_URL'])
        db.prepare('select', PG_QUERY)
        db
      end
    end
  end
end

File: frameworks/symfony-spawn-franken/src/Controller/BenchmarkController.php
<?php

declare(strict_types=1);

namespace App\Controller;

use Doctrine\DBAL\Connection;
use Doctrine\DBAL\ParameterType;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Attribute\Route;

class BenchmarkController
{
    private static array $dataset = [];
    private static array $staticFiles = [];
    private static bool $dataLoaded = false;

    private const MIME_TYPES = [
        'css'   => 'text/css',
        'js'    => 'application/javascript',
        'html'  => 'text/html',
        'woff2' => 'font/woff2',
        'svg'   => 'image/svg+xml',
        'webp'  => 'image/webp',
        'json'  => 'application/json',
    ];

    public function __construct(private readonly Connection $connection)
    {
        if (self::$dataLoaded) {
            return;
        }

        self::$dataset = json_decode(file_get_contents('/data/dataset.json'), true);

        $dir = '/data/static';
        if (is_dir($dir)) {
            foreach (scandir($dir) as $file) {
                if ($file === '.' || $file === '..') continue;
                if (str_ends_with($file, '.br') || str_ends_with($file, '.gz')) continue;
                $base = $dir . '/' . $file;
                $ext  = pathinfo($file, PATHINFO_EXTENSION);
                self::$staticFiles[$file] = [
                    'data' => file_get_contents($base),
                    'mime' => self::MIME_TYPES[$ext] ?? 'application/octet-stream',
                    'br'   => file_exists($base . '.br') ? file_get_contents($base . '.br') : null,
                    'gz'   => file_exists($base . '.gz') ? file_get_contents($base . '.gz') : null,
                ];
            }
        }

        self::$dataLoaded = true;
    }

    #[Route('/baseline11', methods: ['GET', 'POST'])]
    #[Route('/baseline2', methods: ['GET', 'POST'])]
    public function baseline(Request $request): Response
    {
        $sum = array_sum($request->query->all());
        if ($request->isMethod('POST')) {
            $sum += (int) $request->getContent();
        }
        return new Response((string) $sum, 200, ['Content-Type' => 'text/plain']);
    }

    #[Route('/pipeline')]
    public function pipeline(): Response
    {
        return new Response('ok', 200, ['Content-Type' => 'text/plain']);
    }

    #[Route('/json/{count}', requirements: ['count' => '\d+'])]
    public function json(int $count, Request $request): Response
    {
        $count = max(0, min($count, count(self::$dataset)));
        $m = (int) ($request->query->get('m', 1) ?: 1);
        $items = [];
        for ($i = 0; $i < $count; $i++) {
            $item          = self::$dataset[$i];
            $item['total'] = $item['price'] * $item['quantity'] * $m;
            $items[]       = $item;
        }
        return new Response(
            json_encode(['items' => $items, 'count' => $count], JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES),
            200,
            ['Content-Type' => 'application/json']
        );
    }

    #[Route('/upload', methods: ['POST'])]
    public function upload(Request $request): Response
    {
        return new Response((string) strlen($request->getContent()), 200, ['Content-Type' => 'text/plain']);
    }

    #[Route('/async-db')]
    public function asyncDb(Request $request): Response
    {
        $min   = (int) ($request->query->get('min', 10));
        $max   = (int) ($request->query->get('max', 50));
        $limit = max(1, min(50, (int) ($request->query->get('limit', 50))));

        try {
            $stmt = $this->connection->prepare(
                'SELECT id, name, category, price, quantity, active, tags, rating_score, rating_count FROM items WHERE price BETWEEN ? AND ? LIMIT ?'
            );
            $stmt->bindValue(1, $min);
            $stmt->bindValue(2, $max);
            $stmt->bindValue(3, $limit, ParameterType::INTEGER);
            $result = $stmt->executeQuery();
            $rows   = $result->fetchAllAssociative();

            $items = array_map(static function (array $row): array {
                $row['active'] = (bool) $row['active'];
                $row['tags']   = json_decode($row['tags'], true);
                $row['rating'] = [
                    'score' => (int) $row['rating_score'],
                    'count' => (int) $row['rating_count'],
                ];
                unset($row['rating_score'], $row['rating_count']);
                return $row;
            }, $rows);

            return new Response(
                json_encode(['items' => $items, 'count' => count($items)]),
                200,
                ['Content-Type' => 'application/json']
            );
        } catch (\Throwable) {
            return new Response('{"items":[],"count":0}', 200, ['Content-Type' => 'application/json']);
        }
    }

    #[Route('/static/{file}', requirements: ['file' => '.+'])]
    public function static(string $file, Request $request): Response
    {
        if (!isset(self::$staticFiles[$file])) {
            return new Response('Not Found', 404, ['Content-Type' => 'text/plain']);
        }

        $f       = self::$staticFiles[$file];
        $ae      = $request->headers->get('Accept-Encoding', '');
        $headers = ['Content-Type' => $f['mime']];

        if ($f['br'] !== null && str_contains($ae, 'br')) {
            $headers['Content-Encoding'] = 'br';
            return new Response($f['br'], 200, $headers);
        }

        if ($f['gz'] !== null && str_contains($ae, 'gzip')) {
            $headers['Content-Encoding'] = 'gzip';
            return new Response($f['gz'], 200, $headers);
        }

        return new Response($f['data'], 200, $headers);
    }
}
